hoomd/mpcd/ReverseNonequilibriumShearFlow.cc

```diff
@@ -0,0 +1,327 @@
+    detail::ReverseNonequilibriumShearFlowSorter cmp(m_p_target);
+    std::map<Scalar, unsigned int, detail::ReverseNonequilibriumShearFlowSorter>
+        particles_in_lo_slab(cmp);
+    std::map<Scalar, unsigned int, detail::ReverseNonequilibriumShearFlowSorter>
+        particles_in_hi_slab(cmp);
+
+    // sort particles into their slab in z-direction and record momentum in x-direction and index
+    // injecting into std::map will sort them from lowest to highest momentum
+    for (unsigned int j = 0; j < m_mpcd_pdata->getN(); j++)
+        {
+        const Scalar4 vel = h_vel.data[j];
+        const Scalar momentum = vel.x * m_mpcd_pdata->getMass();
+        const Scalar z = h_pos.data[j].z;
+        if (m_lo_pos.x <= z && z < m_lo_pos.y
+            && momentum > 0) // lower slab, search for max momentum
+            {
+            particles_in_lo_slab.insert(std::make_pair(momentum, j));
+            }
+        else if (m_hi_pos.x <= z && z < m_hi_pos.y
+                 && momentum < 0) // higher slab, search for min momentum
+            {
+            particles_in_hi_slab.insert(std::make_pair(momentum, j));
+            }
+        }
+
+    // find the fastest particles (in +x) in the lo slab
+    ArrayHandle<Scalar2> h_layer_lo(m_layer_lo, access_location::host, access_mode::overwrite);
+    m_num_lo = 0;
+    for (auto iter = particles_in_lo_slab.rbegin(); iter != particles_in_lo_slab.rend(); ++iter)
+        {
+        if (m_num_lo < m_num_swap)
+            {
+            // (x,y) = (index, momentum)
+            h_layer_lo.data[m_num_lo++] = make_scalar2(__int_as_scalar(iter->second), iter->first);
+            }
+        }
+
+    // find the slowest particles (in -x) in the high slab
+    ArrayHandle<Scalar2> h_layer_hi(m_layer_hi, access_location::host, access_mode::overwrite);
+    m_num_hi = 0;
+    for (auto iter = particles_in_hi_slab.begin(); iter != particles_in_hi_slab.end(); ++iter)
+        {
+        if (m_num_hi < m_num_swap)
+            {
+            h_layer_hi.data[m_num_hi++] = make_scalar2(__int_as_scalar(iter->second), iter->first);
+            }
+        }
+    }
+
+/*!
+ * Takes the up to m_num_swap pairs from the top and bottom slab and swaps
+ * their velocities.
+ */
+void mpcd::ReverseNonequilibriumShearFlow::swapPairMomentum()
+    {
+    // get all the needed array handles. Only velocities needs to be readwrite
+    ArrayHandle<Scalar4> h_vel(m_mpcd_pdata->getVelocities(),
+                               access_location::host,
+                               access_mode::readwrite);
+    ArrayHandle<Scalar2> h_layer_lo(m_layer_lo, access_location::host, access_mode::read);
+    ArrayHandle<Scalar2> h_layer_hi(m_layer_hi, access_location::host, access_mode::read);
+
+    // find out how many pairs there are
+    unsigned int num_pairs = std::min(m_num_swap, std::min(m_num_lo, m_num_hi));
+
+    // swap num_pairs (up to m_num_swap)
+    m_momentum_exchange = 0;
```

Comment:
If we switch to summed momentum, remove this line
```suggestion
```

Reply:
What of line 152?

Reply:
Yeah, that line also needs to be removed.